packages/optimizer/lib/transformers/OptimizeImages.js

```diff
@@ -0,0 +1,249 @@
+   */
+  setBaseWidth(imgSrcWidth, maxImgWidth = -1) {
+    this.widthList_.length = 0;
+    let previousWidth = -1;
+    if (maxImgWidth > 0 && imgSrcWidth > maxImgWidth) {
+      return;
+    }
+
+    for (let i = NUM_SRCSET_DPR.length - 1; i >= 0; --i) {
+      let width = this.roundUp(Math.ceil(imgSrcWidth * NUM_SRCSET_DPR[i]));
+      if (maxImgWidth > 0 && width > maxImgWidth) {
+        width = maxImgWidth;
+      }
+      if (width != previousWidth) {
+        this.widthList_.push(width);
+      }
+
+      previousWidth = width;
+    }
+  }
+
+  /**
+   *  Returns true if there is more legitimate width.
+   */
+  moreWidth() {
+    return this.widthList_.length > 0;
+  }
+
+  /**
+   * Returns the current legitimate width and moves the state to the next one.
+   */
+  nextWidth() {
+    const nextWidth = this.widthList_[this.widthList_.length - 1];
+    this.widthList_.pop();
+    return nextWidth;
+  }
+
+  /**
+   *
+   */
+  isValid() {
+    return this.widthList_.length > 1;
+  }
+
+  roundUp(value) {
+    for (const width of SRCSET_WIDTH) {
+      if (width > value) {
+        return width;
+      }
+    }
+    return SRCSET_WIDTH[SRCSET_WIDTH.length - 1];
+  }
+}
+
+/**
+ * ImageTransformer - generates srcset attribute for amp-img.
+ *
+ * This transformer supports the following option:
+ *
+ * - `optimizeImages`: set to `true` to enable image optimization, by default it will encode addition widt
+ * - `imageOptimizer`: a function for customizing the srcset generation. The function should return a URL
+ *    pointing to a version of the `src` image with the given `width`. If no image is available, it should
+ *    return a falsy value. For example:
+ *
+ *       (src, width) => `${src}?width=${width}`
+ */
+class OptimizeImages {
+  constructor(config) {
+    this.log = config.log;
+    this.enabled = !!config.optimizeImages;
+    this.imageOptimizer = config.imageOptimizer || DEFAULT_IMAGE_OPTIMIZER;
+    this.srcsetWidth = new SrcsetWidth();
+  }
+
+  async transform(root) {
+    if (!this.enabled) {
+      return;
+    }
+    const html = firstChildByTag(root, 'html');
+    const body = firstChildByTag(html, 'body');
+
+    let node = body;
+    while (node !== null) {
+      if (node.tagName === 'template') {
+        node = skipNodeAndChildren(node);
+      } else {
+        if (node.tagName === 'amp-img') {
+          await this.optimizeImage(node);
+        }
+        node = nextNode(node);
+      }
+    }
+  }
+
+  async optimizeImage(imageNode) {
+    // Don't change existing srcsets.
+    if (hasAttribute(imageNode, 'srcset')) {
+      return;
+    }
+    // Should not happen for valid AMP.
+    if (!hasAttribute(imageNode, 'src')) {
+      return;
+    }
+    const src = imageNode.attribs.src;
+    // Check if it's a relative path or a valid http(s) URL.
+    if (!isValidImageSrcURL(src)) {
+      return;
+    }
+
+    // No srcset is added if the image ends with a `,` (comma). See
+    // http://b/127535381 for context.
```

Comment:
is it ok reference corplinks on a public repo?

Reply:
yeah, it's not nice, but better than leaving it completely undocumented. It's a common pattern in the AMP repo.